File: cmd/pdxl/parse.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/spf13/cobra"
	v3 "pdxl/internal/parser/v3"
)

var parseJSON bool
var parseTree bool

var parseCmd = &cobra.Command{
	Use:   "parse <file>",
	Short: "Parse a Paradox scripting file and print the AST",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		filename := args[0]
		data, err := os.ReadFile(filename)
		if err != nil {
			return fmt.Errorf("reading %s: %w", filename, err)
		}
		tree, diags := v3.Parse(filename, data)
		for _, d := range diags {
			fmt.Fprintf(os.Stderr, "%s\n", d.String())
		}
		switch {
		case parseJSON:
			enc := json.NewEncoder(os.Stdout)
			enc.SetIndent("", "  ")
			return enc.Encode(tree)
		case parseTree:
			printNodeTree(tree)
		default:
			printFlat(tree, 0)
		}
		return nil
	},
}

func init() {
	parseCmd.Flags().BoolVar(&parseJSON, "json", false, "output AST as JSON")
	parseCmd.Flags().BoolVar(&parseTree, "tree", false, "output AST as a labelled node tree")
	rootCmd.AddCommand(parseCmd)
}

// ── flat printer (default) ────────────────────────────────────────────────────

func printFlat(tree *v3.Tree, depth int) {
	root := tree.Root()
	for _, idx := range tree.ChildRefs(root) {
		printFlatNode(tree, tree.Nodes[idx], depth)
	}
}

func printFlatNode(tree *v3.Tree, n v3.Node, depth int) {
	ind := strings.Repeat("\t", depth)
	switch n.Kind {
	case v3.KindField:
		children := tree.Children(n)
		key := children[0].Value(tree.Src)
		op := n.OpString()
		val := children[1]
		switch val.Kind {
		case v3.KindScalar:
			fmt.Printf("%s%s %s %s\n", ind, key, op, val.Value(tree.Src))
		case v3.KindTaggedBlock:
			fmt.Printf("%s%s %s %s {\n", ind, key, op, val.Value(tree.Src))
			for _, idx := range tree.ChildRefs(val) {
				printFlatNode(tree, tree.Nodes[idx], depth+1)
			}
			fmt.Printf("%s}\n", ind)
		case v3.KindBlock:
			fmt.Printf("%s%s %s {\n", ind, key, op)
			for _, idx := range tree.ChildRefs(val) {
				printFlatNode(tree, tree.Nodes[idx], depth+1)
			}
			fmt.Printf("%s}\n", ind)
		}
	case v3.KindScalar:
		fmt.Printf("%s%s\n", ind, n.Value(tree.Src))
	case v3.KindBlock:
		fmt.Printf("%s{\n", ind)
		for _, idx := range tree.ChildRefs(n) {
			printFlatNode(tree, tree.Nodes[idx], depth+1)
		}
		fmt.Printf("%s}\n", ind)
	}
}

// ── tree printer (--tree) ─────────────────────────────────────────────────────

func printNodeTree(tree *v3.Tree) {
	root := tree.Root()
	fmt.Println("Root (KindFile)")
	refs := tree.ChildRefs(root)
	for i, idx := range refs {
		last := i == len(refs)-1
		printTreeNode(tree, tree.Nodes[idx], "", last)
	}
}

func printTreeNode(tree *v3.Tree, n v3.Node, prefix string, last bool) {
	branch := "├── "
	childPrefix := prefix + "│   "
	if last {
		branch = "└── "
		childPrefix = prefix + "    "
	}

	switch n.Kind {
	case v3.KindScalar:
		fmt.Printf("%s%sKindScalar  %q\n", prefix, branch, n.Value(tree.Src))

	case v3.KindField:
		children := tree.Children(n)
		key := children[0].Value(tree.Src)
		op := n.OpString()
		val := children[1]
		fmt.Printf("%s%sKindField   key=%q  op=%q\n", prefix, branch, key, op)
		printTreeNode(tree, val, childPrefix, true)

	case v3.KindBlock:
		fmt.Printf("%s%sKindBlock\n", prefix, branch)
		refs := tree.ChildRefs(n)
		for i, idx := range refs {
			printTreeNode(tree, tree.Nodes[idx], childPrefix, i == len(refs)-1)
		}

	case v3.KindTaggedBlock:
		fmt.Printf("%s%sKindTaggedBlock  tag=%q\n", prefix, branch, n.Value(tree.Src))
		refs := tree.ChildRefs(n)
		for i, idx := range refs {
			printTreeNode(tree, tree.Nodes[idx], childPrefix, i == len(refs)-1)
		}
	}
}

File: rust/crates/pdxl-cli/src/parse.rs
//! `pdxl parse` — print the AST, mirroring Go's `cmd/pdxl/parse.go`:
//! the default flat printer (tab-indented source-like rendering) and the
//! `--tree` labelled node tree with box-drawing characters. Parse diagnostics
//! go to stderr as `file:line:col: message`, exactly like `Diagnostic.String()`.

use std::io::{self, Write};
use std::process::ExitCode;

use pdxl_parser::{NodeId, NodeKind, SyntaxTree, parse};

pub fn run(file: &str, tree_mode: bool) -> io::Result<ExitCode> {
    let data = std::fs::read(file)
        .map_err(|e| io::Error::new(e.kind(), format!("reading {file}: {e}")))?;
    let parsed = parse(file.to_string(), data.clone());
    for d in parsed.diagnostics() {
        let (line, col) = pdxl_src::line_col(&data, d.offset);
        eprintln!("{}:{line}:{col}: {}", d.filename, d.message);
    }

    let stdout = io::stdout();
    let mut w = io::BufWriter::new(stdout.lock());
    let tree = parsed.tree();
    if tree_mode {
        print_node_tree(&mut w, tree)?;
    } else {
        for child in tree.children(tree.root()) {
            print_flat_node(&mut w, tree, child, 0)?;
        }
    }
    w.flush()?;
    Ok(ExitCode::SUCCESS)
}

fn text<'t>(tree: &'t SyntaxTree, id: NodeId) -> std::borrow::Cow<'t, str> {
    String::from_utf8_lossy(tree.node_text(id))
}

// ── flat printer (default) — Go printFlat/printFlatNode ─────────────────────

fn print_flat_node(
    w: &mut impl Write,
    tree: &SyntaxTree,
    id: NodeId,
    depth: usize,
) -> io::Result<()> {
    let ind = "\t".repeat(depth);
    let node = tree.node(id);
    match node.kind {
        NodeKind::Field => {
            let kids = tree.child_ids(id);
            let key = text(tree, kids[0]);
            let op = node.op_string();
            let val = kids[1];
            match tree.node(val).kind {
                NodeKind::Scalar => {
                    writeln!(w, "{ind}{key} {op} {}", text(tree, val))?;
                }
                NodeKind::TaggedBlock => {
                    writeln!(w, "{ind}{key} {op} {} {{", text(tree, val))?;
                    for item in tree.children(val) {
                        print_flat_node(w, tree, item, depth + 1)?;
                    }
                    writeln!(w, "{ind}}}")?;
                }
                NodeKind::Block => {
                    writeln!(w, "{ind}{key} {op} {{")?;
                    for item in tree.children(val) {
                        print_flat_node(w, tree, item, depth + 1)?;
                    }
                    writeln!(w, "{ind}}}")?;
                }
                _ => {}
            }
        }
        NodeKind::Scalar => writeln!(w, "{ind}{}", text(tree, id))?,
        NodeKind::Block => {
            writeln!(w, "{ind}{{")?;
            for item in tree.children(id) {
                print_flat_node(w, tree, item, depth + 1)?;
            }
            writeln!(w, "{ind}}}")?;
        }
        _ => {}
    }
    Ok(())
}

// ── tree printer (--tree) — Go printNodeTree/printTreeNode ──────────────────

fn print_node_tree(w: &mut impl Write, tree: &SyntaxTree) -> io::Result<()> {
    writeln!(w, "Root (KindFile)")?;
    let refs = tree.child_ids(tree.root());
    for (i, id) in refs.iter().enumerate() {
        print_tree_node(w, tree, *id, "", i == refs.len() - 1)?;
    }
    Ok(())
}

fn print_tree_node(
    w: &mut impl Write,
    tree: &SyntaxTree,
    id: NodeId,
    prefix: &str,
    last: bool,
) -> io::Result<()> {
    let (branch, child_prefix) = if last {
        ("└── ", format!("{prefix}    "))
    } else {
        ("├── ", format!("{prefix}│   "))
    };

    let node = tree.node(id);
    match node.kind {
        NodeKind::Scalar => {
            writeln!(w, "{prefix}{branch}KindScalar  {:?}", text(tree, id))?;
        }
        NodeKind::Field => {
            let kids = tree.child_ids(id);
            writeln!(
                w,
                "{prefix}{branch}KindField   key={:?}  op={:?}",
                text(tree, kids[0]),
                node.op_string()
            )?;
            print_tree_node(w, tree, kids[1], &child_prefix, true)?;
        }
        NodeKind::Block => {
            writeln!(w, "{prefix}{branch}KindBlock")?;
            let refs = tree.child_ids(id);
            for (i, kid) in refs.iter().enumerate() {
                print_tree_node(w, tree, *kid, &child_prefix, i == refs.len() - 1)?;
            }
        }
        NodeKind::TaggedBlock => {
            writeln!(
                w,
                "{prefix}{branch}KindTaggedBlock  tag={:?}",
                text(tree, id)
            )?;
            let refs = tree.child_ids(id);
            for (i, kid) in refs.iter().enumerate() {
                print_tree_node(w, tree, *kid, &child_prefix, i == refs.len() - 1)?;
            }
        }
        _ => {}
    }
    Ok(())
}
